api/database/migrations/2022_05_04_173542_add_extra_constraints.php

```diff
@@ -0,0 +1,140 @@
+<?php
+
+use Illuminate\Database\Migrations\Migration;
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Support\Facades\Schema;
+use Illuminate\Support\Facades\DB;
+
+class AddExtraConstraints extends Migration
+{
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        /** the first batch is a bunch of pivots to make unique **/
+
+        // a pool may not have duplicate CMO assets assigned to it
+        Schema::table('asset_cmo_asset_pool', function (Blueprint $table) {
+            $table->unique(['cmo_asset_id', 'pool_id'], 'asset_cmo_asset_pool_unique');
```

Comment:
Does the order of constraints in these pairs matter to us?

I see that sometimes the parent entity is first in the constraint, and sometimes it's second. Not sure if that's intentional. If it is, maybe you could leave comment as to why? And if not, maybe we could keep it consistent to keep future behaviour more consistent? (e.g., partial indexes: https://stackoverflow.com/a/52619215)

Reply:
Sorry, I think I may have mislead you with the description comment.  I'm not sure there's really a "parent" and a "child" in these relationships.  These are all pivot tables to support many-to-many relationships between two peer models.   My comment says "a pool has CMO assets" because that's the way a user would see it.  From a data modelling perspective it would be just as valid to say that "a CMO asset has pools" though it's not a useful use case and our API doesn't have any queries written to support that.

The idea of partial index usage is interesting.  These indexes weren't added for performance but in all these relationships there is one side that we would typically use for searching and filtering.  I just paired them in the order that the columns appear in the tables.
![image](https://user-images.githubusercontent.com/8978655/167635255-b2d7ce14-dd5e-4cb8-b168-e4a876cb3489.png)
I'll switch them to have the "filterable" field first.  Thanks for the idea!